internal/command/arguments/extended.go

```diff
@@ -128,28 +133,97 @@ func parseTargetables(rawTargetables []string, flag string) ([]addrs.Targetable,
 	return targetables, diags
 }
 
-func parseRawTargetsAndExcludes(targets []string, excludes []string) ([]addrs.Targetable, []addrs.Targetable, tfdiags.Diagnostics) {
-	var parsedTargets []addrs.Targetable
-	var parsedExcludes []addrs.Targetable
+// parseFile gets a filePath and reads the file, which contains a list of targets
+// with each line in the file representating a targeted object, and returns
+// a list of addrs.Targetable. This is used for parsing the input of -target-file
+// and -exclude-file flags
+func parseFileTargetables(filePath, flag string) ([]addrs.Targetable, tfdiags.Diagnostics) {
+
+	// If no file passed, no targets
+	if filePath == "" {
+		return nil, nil
+	}
+	var targetables []addrs.Targetable
 	var diags tfdiags.Diagnostics
 
-	if len(targets) > 0 && len(excludes) > 0 {
+	b, err := os.ReadFile(filePath)
+	diags = diags.Append(err)
+
+	sc := hcl.NewRangeScanner(b, filePath, bufio.ScanLines)
+	for sc.Scan() {
+		lineBytes := sc.Bytes()
+		lineRange := sc.Range()
+		if isComment(lineBytes) {
+			continue
+		}
+		traversal, syntaxDiags := hclsyntax.ParseTraversalAbs(lineBytes, lineRange.Filename, lineRange.Start)
+		if syntaxDiags.HasErrors() {
+			diags = diags.Append(nil).Append(
+				&hcl.Diagnostic{
+					Severity: tfdiags.Error.ToHCL(),
+					Summary:  "Invalid syntax",
+					Detail:   fmt.Sprintf("For %s %q: %v", flag, lineBytes, syntaxDiags[0].Detail),
+					Subject: &hcl.Range{
+						Filename: lineRange.Filename,
+						Start:    lineRange.Start,
+						End:      lineRange.End,
+					},
+				},
+			)
+			continue
+		}
+
+		target, targetDiags := addrs.ParseTarget(traversal)
+		if targetDiags.HasErrors() {
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				fmt.Sprintf("Invalid %s address", flag),
+				fmt.Sprintf("my detail: %v %s %q", targetDiags[0].Description().Detail, flag, lineBytes),
+			))
+			continue
+		}
+
+		targetables = append(targetables, target.Subject)
+	}
+	return targetables, diags
+}
+
+func isComment(b []byte) bool {
+	if strings.HasPrefix(string(b), "#") {
+		return true
+	}
+	return false
+}
+
+func parseRawTargetsAndExcludes(targetsDirect, excludesDirect []string, targetFile, excludeFile string) ([]addrs.Targetable, []addrs.Targetable, tfdiags.Diagnostics) {
```

Comment:
One thing I noticed looking at this signature is that it only accepts a single file in each case. I suppose this is okay, but it's a little inconsistent with `-var-file` where we allow any mix of any number of `-var` and `-var-file` options on the command line.

I expect we could extend this to support multiple `-target-file` or `-exclude-file` options later without any backward-compatibility problems and so starting with the simpler case is probably fine... but I figured I'd raise it now in case you think that the more flexible design is straightforward enough to implement immediately.

Reply:
This is an awesome comment. I can most definitely work on this and incorporate multiple files into the feature. Thank you for the prelim review. If you see more things, please comment away when you have time, and I’ll just keep slugging away until you and the rest of the team are happy with the code.
Being my first PR to the codebase, I want to make sure I respect the quality of the codebase.